### PR TITLE
feat(replica): replace slaves by replica while keeping backward comp

### LINKED ISSUE #154: `add_slave` should be deprecated in favor of `add_replica`
We should stop using the word `slave` and use `replica` instead. The change is already propagated in the API, we now need to update our public methods.

I think `add_slave` should be renamed to `add_replica`, and any mention of the method in the documentation should be updated.

We should keep `add_slave` as an alias for `add_replica` for now, but throw a deprecation warning when used. It will be removed in a future major version.

### PR DESCRIPTION
**Pull request by Jerskouille**
@alexandremeunier had started it, but I think to keep the backward comp we can't just do a `sed` here.

Closes #154 .

### CODE REVIEW

**Comment by Jerskouille** [comment]
Should be good now @redox !

File: lib/algoliasearch-rails.rb [hunk 1]
```diff
@@ -225,28 +225,29 @@ def to_settings
...
 
-    def add_slave(index_name, options = {}, &block)
-      raise ArgumentError.new('Cannot specify additional slaves on a slave index') if @options[:slave]
+    def add_replica(index_name, options = {}, &block)
+      raise ArgumentError.new('Cannot specify additional replicas on a replica index') if @options[:slave] || @options[:replica]
       raise ArgumentError.new('No block given') if !block_given?
-      add_index(index_name, options.merge({ :slave => true }), &block)
+      add_index(index_name, options.merge({ :replica => true }), &block)
     end
+    alias :add_slave :add_replica
```
Comment (pixelastic) [inline lib/algoliasearch-rails.rb]: `add_slave` should be a bit more than an alias and throw a deprecation warning when used

**Comment by pixelastic** [comment]
Also, we need to be careful about the way the API handles the backward compat and need to be compliant with it.

If you already have defined a slave in the past using `set_settings`, you will get a `:slave` key  in your `get_settings`. When you then call `add_replica`, it should not push a new `:replicas` key, but it should update the existing `:slaves` key.

In other words we want the Rails API to only ever mention `replicas`, *but* if users already had `slaves` defined, we need to make sure they won't lose their config. The name of the key that should be used to store this info in the index config should be the key that was already used before (if any), or fallback to `replicas`.

**Comment by Jerskouille** [comment]
Valid concerns @pixelastic .

File: lib/algoliasearch-rails.rb (hunk 1 above)
Comment (redox) [inline lib/algoliasearch-rails.rb]: This is not the way we went so far; I don't think we want to print something in the logs for instance.

Comment (pixelastic) [inline lib/algoliasearch-rails.rb]: How would you tell the user that we will deprecate the feature then? You want to keep it an undocumented silent alias?

Comment (Jerskouille) [inline lib/algoliasearch-rails.rb]: Breaking change on new major release ?

Comment (pixelastic) [inline lib/algoliasearch-rails.rb]: With no warning beforehand? How do we usually handle deprecation of features?

Comment (Jerskouille) [inline lib/algoliasearch-rails.rb]: I think we don't. :)
Have we ever had another feature deprecation ?

**Comment by Jerskouille** [comment]
Since we're not doing a `get_settings` to set the settings of the index, we can't know which key the API is currently using.
We instead use the `algoliasearch` block methods to set the settings, so this exact behavior cannot be reproduced reliably.

However, we could do something simple:
- `add_slave` sets the `slaves` setting
- `add_replica` sets the `replicas` setting
- `raise` on mixed usage of `add_slave` and `add_replica`

WDYT @pixelastic @redox ?

**Comment by redox** [comment]
> Since we're not doing a get_settings to set the settings of the index, we can't know which key the API is currently using.

Actually we do; but I'm just wondering: what if we always set `replicas` to the list of indices and `slaves` to `null` -> that will work no? (in rails, we can safely assume that only that block code is updating the settings).

**Comment by Jerskouille** [comment]
The main concern I see here : it might break any other script targeting the same index depending on `.get_settings['slaves']` that would start receiving null instead of the list.

**Comment by Jerskouille** [comment]
I've pushed a commit implemented what I talked about above ( see the tests here : https://github.com/algolia/algoliasearch-rails/pull/155/files#diff-078c1c61de993e867792139abaaba7d3R936 )

**Comment by redox** [comment]
@Jerskouille did you see that 1.8.7 failed again?

**Comment by Jerskouille** [comment]
Yes, on it, I'm ultra weirded out by this issue.

**Comment by Jerskouille** [comment]
Wow, just wow.
Wonderful scoping by Ruby 1.8.7 !

```bash
rvm use 1.8.7 && echo '
a = 1
b = [2]
b.any? { |a| }
puts a' | ruby
```

Guess the output ?

**Comment by pixelastic** [comment]
How to solve the backward compat while still giving great DX is really head scratching. For context, here are how things are currently working in other parts of the Algolia landscape:

## In the API

If you send both `slaves` and `replicas` when updating your settings, the call will be rejected. You have to pick one. Whichever you pick has no incidence on the feature, but this will be the name of the key you'll get on subsequent `getSettings` calls. If you send `slaves`, you'll have a `slaves` key in your settings, if you send `replicas`, you'll have a `replicas` key. The other key will be dropped. Only one of the two can ever be defined in the settings.

## In the Dashboard

When you create an index, the settings do not hold neither a `slaves` nor a `replicas` key. If you go to the Dashboard to manually add some, it will create the key at that time. Old users had it saved in `slaves` and new users have it in `replicas`.

When you then update the list of Replicas through the Dashboard, it will be clever and update the key that already exists. It will create one only if none were set before. Note that clearing the list of replicas and then re-adding them will **not** rename the key.

## What to do now?

In order to stay consistent with our current live implementations while still provide backward compat and great DX, here is what I suggest:

For API clients that blindly push a settings object with `setSettings`, we do not need to add anything. We could check that the object does not contain both `slaves` and `replicas` and warn the user, but the API will reject the call anyway with an error message. The current error message is not very clear, but we're working on improving it.

For deeper framework and integrations, that will do a merge of the current settings with the updated ones, we cannot just add a `replicas` key in the mix as the call will get rejected if there already is a `slaves` one. Even with a clear error message from the API, the user will have trouble understanding the issue and fixing it. In that case, we should re-use the Dashboard logic and re-use the existing key.

We have the issue for Rails today, but I know we have a similar issue in Magento (in regards to `attributesToIndex`/`searchableAttributes`). We should try to find the best way to fix it and apply it to all our impacted applications.

_Back to the issue at hand..._

If we want to follow the Dashboard logic here, we would need to know which key is already set. Apparently, this is an information we do not have when we do `add_replicas`/`add_slaves` but only when we merge current settings with new settings right before sending the request.

To keep the same behavior, what we could do is have `add_slaves` and `add_replicas` update their respective keys like in the current commit. But update the setting merge logic to re-use an existing key is one is already set. Would that be possible?

**Comment by redox** [comment]
> Guess the output ?

Arg /o\ `2` ?

**Comment by Jerskouille** [comment]
Exactly !
OK, so what do we do now?

The last commit should do what @pixelastic asked for, i.e. reuse the key we get in our `get_settings`. Do we keep it and go forward? :)

**Comment by redox** [comment]
> The last commit should do what @pixelastic asked for, i.e. reuse the key we get in our get_settings. Do we keep it and go forward? :)

I think we're good @Jerskouille ; awesome (and painful) job !

**Comment by pixelastic** [comment]
Thaaaanks :tada:

**Comment by Jerskouille** [comment]
To clarify, I've just realized that I implemented it as first proposed by @pixelastic .
To clarify, there were 2 propositions:
1. always use `replicas` unless the index already has `slaves` (like the dashboard does)
2. if set in the index, use the index key. Otherwise, use `replicas` if `add_replicas` was used and use `slaves` if `add_slaves` was used.

I had written the tests with 2. in mind, but I implemented the code as 1. .

The tests passed because the index wasn't deleted, only cleared, and since it already existed with the `slaves` key, they passed. But as soon as you tried on another app, they didn't work anymore.

The quick win was to force `slaves` to be used at the index level for the `WithSlaves` tests: https://github.com/algolia/algoliasearch-rails/commit/e2c5dfba7a0f792f9c48455885b227ee6f762448 .

**Pull request merged**